validator/client/src/main/java/tech/pegasys/teku/validator/client/AttestationDutyLoader.java

```diff
@@ -81,26 +108,68 @@ protected SafeFuture<Optional<AttesterDuties>> requestDuties(
     final SlotBasedScheduledDuties<AttestationProductionDuty, AggregationDuty> scheduledDuties =
         scheduledDutiesFactory.apply(duties.getDependentRoot());
 
-    final Optional<DvtAttestationAggregations> dvtAttestationAggregationsForEpoch =
+    final Optional<DvtAttestationAggregations> dvtAttestationAggregations =
         useDvtEndpoint
             ? Optional.of(
                 new DvtAttestationAggregations(validatorApiChannel, duties.getDuties().size()))
             : Optional.empty();
 
-    return SafeFuture.allOf(
-            duties.getDuties().stream()
-                .map(
-                    duty ->
-                        scheduleDuties(scheduledDuties, duty, dvtAttestationAggregationsForEpoch))
-                .toArray(SafeFuture[]::new))
+    if (duties.getDuties().size() < MIN_SIZE_TO_BATCH_BY_SLOT) {
+      return scheduleDuties(scheduledDuties, duties.getDuties(), dvtAttestationAggregations)
+          .<SlotBasedScheduledDuties<?, ?>>thenApply(__ -> scheduledDuties)
+          .alwaysRun(beaconCommitteeSubscriptions::sendRequests);
+    }
+
+    final boolean isCurrentEpoch =
+        spec.computeEpochAtSlot(currentSlot.get()).isLessThanOrEqualTo(epoch);
+
+    final Map<UInt64, List<AttesterDuty>> dutiesBySlot =
+        duties.getDuties().stream()
+            .collect(
+                Collectors.groupingBy(AttesterDuty::getSlot, TreeMap::new, Collectors.toList()));
+
+    SafeFuture<Void> dutiesScheduling = SafeFuture.COMPLETE;
+
+    int i = 0;
```

Comment:
i'd say we'll want to break this down a little if we're going to have a solid chance of unit testing...